src/stdlib/gospelstdlib.mli

```diff
@@ -31,359 +31,576 @@
 
-  (*@ function hd (s: 'a seq) : 'a = s[0] *)
-  (*@ function tl (s: 'a seq) : 'a seq = s[1 ..] *)
+  (*@ function hd (l: 'a t) : 'a *)
+  (** [hd l] is the first element of [l]. *)
 
-  (* Sorted sequences of int values *)
-  (*@ predicate sorted_sub (s: int seq) (l u: integer) =
-        forall i1 i2. l <= i1 <= i2 < u -> s[i1] <= s[i2] *)
-  (*@ predicate sorted (s: int seq) =
-        sorted_sub s 0 (length s) *)
+  (*@ function tl (l: 'a t) : 'a t *)
+  (** [tl l] is [x], when [t] is [.. :: t]. *)
 
-  (* hd, tl, rev, mem *)
-  (* higher-order: map, fold, exists, forall, find, partition *)
-  (* assoc, mem_assoc? split, combine? *)
-end
+  (*@ function nth (l: 'a t) (i: integer) : 'a *)
+  (** [nth l i] is the [i]th element of [l]. *)
 
-module SeqPerm : sig
+  (*@ function nth_opt (l: 'a t) (i: integer) : 'a option *)
+  (** [nth l i] is the [i]th element of [l] if [i] is within the bounds of [l],
+      and [None] otherwise. *)
 
-  (*@ function occ (s1: 'a seq) (v: 'a) (l u: integer) : integer *)
+  (*@ function rev (l: 'a t) : 'a t *)
+  (** [rev l] contains the same elements as [l] in a reverse order. *)
 
-  (*@ axiom occ_base: forall s1 v l u.
-        u <= l -> occ s1 v l u = 0 *)
+  (*@ function init (n: integer) (f: integer -> 'a) : 'a t *)
+  (** [init n f] is a list of length [n], with element number [i] initialized
+      with [f i]. *)
 
-  (*@ axiom occ_ind: forall s1 v l u. 0 <= l <= v < u <= length s1 ->
-        occ s1 v l u = (if v = s1[l] then 1 else 0) + occ s1 v (l+1) u *)
+  (*@ function map (f: 'a -> 'b) (l: 'a t) : 'b t *)
+  (** [map f l] applies function [f] to all the elements of [l], and builds a
+      list with the results returned by [f] *)
 
-  (*@ predicate permut (s1 s2: 'a seq) (l u: integer) =
-        length s1 = length s2 &&
-        forall x. occ x s1 l u = occ x s2 l u *)
+  (*@ function mapi (f: integer -> 'a -> 'b) (l: 'a t) : 'b t *)
+  (** Same as {!map}, but the function is applied to the index of the element as
+      first argument, and the element itself as second argument. *)
 
-  (*@ predicate permut_all (s1 s2: 'a seq) =
-        permut s1 s2 0 (length s1) *)
+  (*@ function fold_left (f: 'a -> 'b -> 'a) (init: 'a) (l: 'b t) : 'a *)
+  (** [fold_left f init t] is [f (... (f (f init a\[0\]) a\[1\]) ...) a\[n-1\]],
+      where [n] is the length of [t]. *)
 
-end
+  (*@ function fold_right (f: 'b -> 'a -> 'a) (l: 'b t) (init: 'a) : 'a *)
+  (** [fold_right f t init] is
+      [f a\[0\] (f a\[1\] ( ... (f a\[n-1\] init) ...))], where [n] is the
+      length of [t]. *)
 
-(** Lists
+  (*@ function map2 (f: 'a -> 'b -> 'c) (l: 'a t) (l': 'b t) : 'c t *)
+  (** [map2 f l l'] applies function [f] to all the elements of [l] and [l'],
+      and builds a list with the results returned by [f]. *)
 
-     Type 'a list, [] and (::) constructors are built-in *)
+  (*@ predicate for_all (f: 'a -> bool) (l: 'a t) *)
+  (** [for_all f l] checks if all elements of [l] satisfy the predicate [f]. *)
```

Comment:
The documentation of some predicates/quantifier is written with `checks` and some other with `holds iff`. Maybe we should choose one.

Reply:
`holds iff` is probably better, those are predicates, not functions, so they don't really check anything